programs/marginfi/src/state/price.rs

```diff
@@ -355,11 +391,237 @@ impl PriceAdapter for SwitchboardV2PriceFeed {
     }
 }
 
+fn load_price_update_v2_checked(ai: &AccountInfo) -> MarginfiResult<PriceUpdateV2> {
+    check!(
+        ai.owner.eq(&PYTH_PUSH_ORACLE_ID),
+        MarginfiError::InvalidOracleAccount
+    );
+
+    let price_feed_data = ai.try_borrow_data()?;
+    let discriminator = &price_feed_data[0..8];
+
+    check!(
+        discriminator == <PriceUpdateV2 as anchor_lang_29::Discriminator>::DISCRIMINATOR,
+        MarginfiError::InvalidOracleAccount
+    );
+
+    Ok(PriceUpdateV2::deserialize(
+        &mut &price_feed_data.as_ref()[8..],
+    )?)
+}
+
+#[cfg_attr(feature = "client", derive(Clone, Debug))]
+pub struct PythPushOraclePriceFeed {
+    ema_price: Box<pyth_solana_receiver_sdk::price_update::Price>,
+    price: Box<pyth_solana_receiver_sdk::price_update::Price>,
+}
+
+impl PythPushOraclePriceFeed {
+    /// Pyth push oracles are update using crosschain messages from pythnet
+    /// There can be multiple pyth push oracles for a given feed_id. Marginfi allows using any
+    /// pyth push oracle with a sufficient verification level and price age.
+    ///
+    /// Meaning that when loading the pyth push oracle, we don't verify the account address
+    /// directly, but rather we verify the feed_id in the oracle data.
+    pub fn load_checked(
+        ai: &AccountInfo,
+        feed_id: &FeedId,
+        clock: &Clock,
+        max_age: u64,
+    ) -> MarginfiResult<Self> {
+        let price_feed_account = load_price_update_v2_checked(ai)?;
+
+        let price = price_feed_account
+            .get_price_no_older_than_with_custom_verification_level(
+                clock,
+                max_age,
+                feed_id,
+                MIN_PYTH_PUSH_VERIFICATION_LEVEL,
+            )
+            .map_err(|e| {
+                debug!("Pyth push oracle error: {:?}", e);
+
+                match e {
+                    pyth_solana_receiver_sdk::error::GetPriceError::PriceTooOld => {
+                        MarginfiError::StaleOracle
+                    }
+                    _ => MarginfiError::InvalidOracleAccount,
+                }
+            })?;
+
+        let ema_price = {
+            let price_update::PriceFeedMessage {
+                exponent,
+                publish_time,
+                ema_price,
+                ema_conf,
+                ..
+            } = price_feed_account.price_message;
+
+            pyth_solana_receiver_sdk::price_update::Price {
+                price: ema_price,
+                conf: ema_conf,
+                exponent,
+                publish_time,
+            }
+        };
+
+        Ok(Self {
+            price: Box::new(price),
+            ema_price: Box::new(ema_price),
+        })
+    }
```

Comment:
Is the purpose of the boxes here to reduce stack usage? Seems like we're still creating two prices on stack before moving them to boxes. There may be better alternatives here for this.

Reply:
The issue with high stack usage happens downstream when these adapters are used in the risk engine.

Reply:
This is just replicating the patterns in the pyth adapter already